staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go

```diff
@@ -6895,15 +6943,262 @@ func TestValidateCustomResourceDefinitionValidation(t *testing.T) {
 			},
 			wantError: false,
 		},
+		{
+			name: "valid x-kubernetes-validations for scalar element",
+			input: apiextensions.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensions.JSONSchemaProps{
+						"subRoot": {
+							Type: "string",
+							XValidations: apiextensions.ValidationRules{
+								{
+									Rule:    "self.startsWith('s')",
+									Message: "subRoot should start with 's'.",
+								},
+								{
+									Rule:    "self.endsWith('s')",
+									Message: "subRoot should end with 's'.",
+								},
+							},
+						},
+					},
+				},
+			},
+			wantError:                  false,
+			setRequireStructuralSchema: true,
+		},
+		{
+			name: "valid x-kubernetes-validations for object",
+			input: apiextensions.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					XValidations: apiextensions.ValidationRules{
+						{
+							Rule:    "minReplicas <= maxReplicas",
+							Message: "minReplicas should be no greater than maxReplicas",
+						},
+					},
+					Properties: map[string]apiextensions.JSONSchemaProps{
+						"minReplicas": {
+							Type: "integer",
+						},
+						"maxReplicas": {
+							Type: "integer",
+						},
+					},
+				},
+			},
+			wantError:                  false,
+			setRequireStructuralSchema: true,
+		},
+		{
+			name: "invalid x-kubernetes-validations with empty rule",
+			input: apiextensions.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					XValidations: apiextensions.ValidationRules{
+						{Message: "empty rule"},
+					},
+				},
+			},
+			wantError:                  true,
+			checkErrorMessage:          true,
+			expectedErrMessage:         "rule is not specified",
+			setRequireStructuralSchema: true,
+		},
+		{
+			name: "valid x-kubernetes-validations with empty validators",
+			input: apiextensions.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type:         "object",
+					XValidations: apiextensions.ValidationRules{},
+				},
+			},
+			wantError:                  false,
+			setRequireStructuralSchema: true,
+		},
+		{
+			name: "invalid rule in x-kubernetes-validations",
```

Comment:
Add some more invalid scenarios. For example:

- unknown property
- non-escaped reserved word (probably need one for valid escaped word too).

You can have them in a single test case with multiple validation rules, so that you test that the function is returning all the errors.

Reply:
Added test cases for unknown property and escaping and we have separate test cases for escaping under staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/escaping_test.go as well.